HPXMLOutputReport/measure.rb

```diff
@@ -0,0 +1,418 @@
+# frozen_string_literal: true
+
+# see the URL below for information on how to write OpenStudio measures
+# http://nrel.github.io/OpenStudio-user-documentation/reference/measure_writing_guide/
+
+require_relative 'resources/constants.rb'
+
+# start the measure
+class HPXMLOutputReport < OpenStudio::Measure::ReportingMeasure
+  # human readable name
+  def name
+    # Measure name should be the title case of the class name.
+    return 'HPXML Output Report'
+  end
+
+  # human readable description
+  def description
+    return 'Reports HPXML outputs for residential HPXML-based models.'
+  end
+
+  # human readable description of modeling approach
+  def modeler_description
+    return 'Parses the HPXML file and reports pre-defined outputs.'
+  end
+
+  # define the arguments that the user will input
+  def arguments(model)
+    args = OpenStudio::Measure::OSArgumentVector.new
+
+    format_chs = OpenStudio::StringVector.new
+    format_chs << 'csv'
+    format_chs << 'json'
+    arg = OpenStudio::Measure::OSArgument::makeChoiceArgument('output_format', format_chs, false)
+    arg.setDisplayName('Output Format')
+    arg.setDescription('The file format of the annual (and timeseries, if requested) outputs.')
+    arg.setDefaultValue('csv')
+    args << arg
+
+    return args
+  end
+
+  # define what happens when the measure is run
+  def run(runner, user_arguments)
+    super(runner, user_arguments)
+
+    model = runner.lastOpenStudioModel
+    if model.empty?
+      runner.registerError('Cannot find OpenStudio model.')
+      return false
+    end
+    model = model.get
+
+    # use the built-in error checking
+    if !runner.validateUserArguments(arguments(model), user_arguments)
+      return false
+    end
+
+    output_format = runner.getStringArgumentValue('output_format', user_arguments)
+
+    sqlFile = runner.lastEnergyPlusSqlFile
+    if sqlFile.empty?
+      runner.registerError('Cannot find EnergyPlus sql file.')
+      return false
+    end
+    sqlFile = sqlFile.get
+    if not sqlFile.connectionOpen
+      runner.registerError('EnergyPlus simulation failed.')
+      return false
+    end
+    model.setSqlFile(sqlFile)
+
+    hpxml_defaults_path = model.getBuilding.additionalProperties.getFeatureAsString('hpxml_defaults_path').get
+    hpxml = HPXML.new(hpxml_path: hpxml_defaults_path)
+
+    # Set paths
+    output_dir = File.dirname(sqlFile.path.to_s)
+    output_path = File.join(output_dir, "hpxml_output.#{output_format}")
```

Comment:
Let's call it "results_hpxml" instead of "hpxml_output", as that will be consistent with the existing "results_annual" and "results_timeseries" output files.